### DISCUSSION TITLE
Velocity of Development

### DISCUSSION THREAD

**Discussion by zkhcohen** [jellyfin/jellyfin-androidtv#2218] [General]
With 165 PRs in the 0.15 release milestone, I'd like to pose the idea of iterating faster.

Are there any current obstacles to releasing at a faster cadence to reduce the number of issues and features in a particular release candidate?

**Comment by Dnkhatri** [comment]
I agree apart from that maybe nightlies should be published after every pr. As that allows for testing and allow people that want to use the latest PR they can.

**Reply by nielsvanvelzen** [reply]
Every branch, commit and pull request in this repository is automatically build and publishes an apk file that you can sideload for testing. I'm running these development builds on my devices for testing.

You can download these by going to the actions tab in GitHub and open the "build" workflow for a particular commit/pr/branch.

**Comment by joshuaboniface** [comment]
Our rule, as a project, is to release slowly and test thoroughly (as much as possible), to ensure we don't break the literally hundreds of thousands of installed users. It will be released when it's ready.

**Reply by zkhcohen** [reply]
I'm sorry if this post came across as condescending or demanding -- it looks like we both work in the DevOps space in our day jobs, so we're both familiar with releasing software without breaking it for our users.

It seems counterintuitive to test dozens of features and hundred(s) of PRs in a single release with such a small team (1.5 very active developers on the androidtv app), rather than adopting a modern DevOps process which aims to test and release in smaller chunks.

If nothing else, having an opt-in beta and/or nightly branch would help identify issues faster.

If we're willing to test something like that, I'd gladly contribute to the effort.

**Reply by joshuaboniface** [reply]
The main issue is that client-facing software like this, as opposed to something on the server side, is far harder to test (testing UX and such across dozens of potential devices is an unsolved problem) and far harder to roll back or hotfix if we make a mistake (it means a new release that people have to get either way, during which time we have a lot of complaining users). So we prefer to be thorough rather than fast.

We do welcome any contributions though, if you have ideas on how to assist there. No offense taken.

**Reply by anthonylavado** [reply]
I should add as there was mention of a beta - there actually _is_ an opt-in beta on Google Play, and when the team decides to cut a release candidate, builds are pushed to that channel. It's not as widely promoted, but I seem to recall there being around 1,500 or so installs on that channel (for Android TV).

See:
https://jellyfin.org/posts/android-betas
https://github.com/jellyfin/jellyfin-androidtv/releases?q=Beta&expanded=true

**Reply by thornbill** [reply]
> test and release in smaller chunks

This part is also particularly problematic for this app. There are major portions of the legacy codebase that quite frankly are hot garbage and need major rewrites. It's not unusual for there to be major breakages in master while a section of code is being rewritten.

**Reply by zkhcohen** [reply]
> there actually _is_ an opt-in beta on Google Play

 Yes, sorry, you're correct. I'm actually opted into that one. The time/effort between RCs seems quite high, however. 

Nightlies would be nice so that users who are willing to test are able to contribute without having to package and transfer the software themselves.

> There are major portions of the legacy codebase that quite frankly are hot garbage and need major rewrites. It's not unusual for there to be major breakages in master while a section of code is being rewritten.

This is an excellent point, but also begs the question of whether breaking changes should be merged into the trunk. Obviously that's not always avoidable, hence the use-case for easily and regularly testable branches.

Maybe this just isn't possible in its current state, but I'd love to see it in the future.

**Reply by nielsvanvelzen** [reply]
> but also begs the question of whether breaking changes should be merged into the trunk.

Yes because managing multiple branches with 1 developer is a no-go.

> hence the use-case for easily and regularly testable branches.

That's the master branch. You can download and sideload builds for testing. Although generally I recommend waiting for beta releases.

**Comment by nielsvanvelzen** [comment]
> With 165 PRs in the 0.15 release milestone, I'd like to pose the idea of iterating faster.

The counter goes up fast when refactoring and with automatic dependency updates. There's 163 PR's in the 0.15.0 milestone, 60 of which are made by Renovate (dependency updates) and 95 of which are mine. I haven't worked on much features, rather I broke the master branch for a while to refactoring a lot of code. Only recently (yesterday) I started to considered the master branch stable enough again for a next beta release.

> Are there any current obstacles to releasing at a faster cadence to reduce the number of issues and features in a particular release candidate?

We previously released at most once a year and this year we'll have 3 big updates, so we're already speeding up. Mind you, pretty much all changes come from one person (me) that works on Jellyfin in his spare time.
